nimare/nimads.py

```diff
@@ -0,0 +1,415 @@
+        self.description = condition["description"]
+        self.weight = weight
+
+    def to_dict(self):
+        """Convert the Condition to a dictionary."""
+        return {"name": self.name, "description": self.description, "weight": self.weight}
+
+
+class Annotation:
+    """A collection of labels and associated weights from the same Annotator.
+
+    Attributes
+    ----------
+    term_weights : :obj:`pandas.DataFrame`
+        A pandas DataFrame containing the annotation group's labels and weights.
+        This is the main attribute of interest for NeuroStore.
+        A dictionary could also work.
+
+    Notes
+    -----
+    Where would p(term|topic) and p(voxel|topic) arrays/DataFrames go? Having one Annotation per
+    Analysis (for each Annotator), and storing these arrays in the Annotation, would make for
+    *a lot* of duplication.
+    The same goes for metadata/provenance, but that will generally be much lighter on memory than
+    the arrays.
+
+    Could be a dictionary with analysis objects as keys?
+    (need to define __hash__ and __eq__ for Analysis)
+    Or could use Analysis.id as key.
+    """
+
+    def __init__(self, source, studyset):
+        self.name = source["name"]
+        self.id = source["id"]
+        self._analysis_ref = {
+            a.id: weakref.proxy(a) for study in studyset.studies for a in study.analyses
+        }
+        self.notes = [Note(self._analysis_ref[n["analysis"]], n["note"]) for n in source["notes"]]
+        for note in self.notes:
+            self._analysis_ref[note.analysis.id].annotations[self.id] = note.note
+
+    def to_dict(self):
+        """Convert the Annotation to a dictionary."""
+        return {"name": self.name, "id": self.id, "notes": [note.to_dict() for note in self.notes]}
+
+
+class Note:
+    """A Note within an annotation.
+
+    Attributes
+    ----------
+    analysis : Analysis object
+        the analysis the note is associated with
+    note : dict
+        the attributes pertaining to the analysis
+    """
+
+    def __init__(self, analysis, note):
+        self.analysis = analysis
+        self.note = note
+
+    def to_dict(self):
+        """Convert the Note to a dictionary."""
+        return {"analysis": self.analysis.id, "note": self.note}
+
+
+class Image:
+    """A single statistical map from an Analysis.
+
+    Attributes
+    ----------
+    filename
+    type?
+
+    Notes
+    -----
+    Should we support remote paths, with some kind of fetching method?
+    """
+
+    def __init__(self, source):
+        self.url = source["url"]
+        self.filename = source["filename"]
+        self.space = source["space"]
+        self.value_type = source["value_type"]
+
+    def to_dict(self):
+        """Convert the Image to a dictionary."""
+        return {
+            "url": self.url,
+            "filename": self.filename,
+            "space": self.space,
+            "value_type": self.value_type,
+        }
+
+
+class Point:
```

Comment:
My only concern here is that this is sort of "hard coded". Should be fine for now but ideally this structure could be autogenerated from somewhere else?

I wonder if there's any way to auto generate an object given a JSON Schema

Reply:
yep: https://docs.pydantic.dev/datamodel_code_generator/.

this can be an issue for a refactor.